go-controller/pkg/ovn/policy.go

```diff
@@ -783,83 +780,130 @@ func (oc *Controller) addNetworkPolicy(policy *knet.NetworkPolicy) {
 			oc.handlePeerService(policy, ingress, np)
 		}
 
-		for _, fromJSON := range ingressJSON.From {
-			// Add IPBlock to ingress network policy
-			if fromJSON.IPBlock != nil {
-				ingress.addIPBlock(fromJSON.IPBlock)
+		selectors := []podset.Selector{}
+
+		// loop through peers.
+		for _, peer := range ingressRule.From {
+			// Peer rules have 4 possible configurations
+			// - IPBlock is set
+			// - Namespace + Name selectors are set - AND together
+			// - Namespace selector is set: consume all pods in that namespace
+			// - Pod selector is set: assume same namespace as NP
+			if peer.IPBlock != nil {
+				ingress.addIPBlock(peer.IPBlock)
+
+			} else if peer.PodSelector != nil && peer.NamespaceSelector != nil {
+				nss, _ := metav1.LabelSelectorAsSelector(peer.NamespaceSelector)
+				ps, _ := metav1.LabelSelectorAsSelector(peer.PodSelector)
+				selectors = append(selectors, podset.Selector{
+					NamespaceSelector: nss,
+					PodSelector:       ps,
+				})
+
+			} else if peer.PodSelector == nil && peer.NamespaceSelector != nil {
+				// Shortcut: we already maintain an address set for namespaces
+				// So, we can just use a simple function to add and remove that set
+				// This is a defer because it needs the lock (which we currently hold), and adding
+				// a handler is a zalgo-style call.
+				defer oc.handlePeerNamespaceSelector(policy, peer.NamespaceSelector, ingress, np)
```

Comment:
Zalgo, in in this case, means the commonly-cited issue in concurrent systems when a callback can be executed either synchronously or asynchronously.

We have this situation here, and it is unfortunate.